src/oni2-sparkle/stubs/winsparkle.c

```diff
@@ -0,0 +1,127 @@
+#include "config.h"
+
+#include <caml/alloc.h>
+#include <caml/callback.h>
+#include <caml/memory.h>
+#include <caml/mlvalues.h>
+#include <caml/threads.h>
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#ifdef USE_WIN_SPARKLE
+
+#include <windows.h>
+#include <shlwapi.h>
+
+#include "winsparkle.h"
+
+void oni2_WinSparkleCloseCallback() {
+  static const value *closeCallback;
+  if (closeCallback == NULL) {
+    closeCallback = caml_named_value("oni2_close");
+  }
+
+  if (closeCallback != NULL) {
+    caml_c_thread_register();
+    caml_acquire_runtime_system();
+    caml_callback(*closeCallback, Val_unit);
+    caml_release_runtime_system();
+  }
+}
+
+CAMLprim value oni2_SparkleInit() {
+  CAMLparam0();
+
+  char iniPath[MAX_PATH];
+
+  GetModuleFileName(NULL, iniPath, MAX_PATH);
+  PathRemoveFileSpec(iniPath);
+
+  strcat(iniPath, "\\Oni2.ini");
+  char version[16];
+  wchar_t versionWide[16];
+  
+
+  GetPrivateProfileString("Application", "Version", "", version, 16, iniPath);
+
+  mbstowcs(versionWide, version, 16);
+
+  win_sparkle_set_app_details(L"Outrun Labs LLC", L"Onivim 2", versionWide);
+  win_sparkle_set_shutdown_request_callback(oni2_WinSparkleCloseCallback);
+  win_sparkle_init();
+
+  CAMLreturn(Val_unit);
+}
+
+// On WinSparkle, there is no instance -- the library just uses global variables.
+// So here we just return a basic `unit`
+CAMLprim value oni2_SparkleGetSharedInstance() {
+  CAMLparam0();
+
+  CAMLreturn(Val_unit);
+}
+
+// This function doesn't make too much sense since we lose the OOP aspect on
+// Windows, so it's basically a noop.
+CAMLprim value oni2_SparkleDebugToString(value vData) {
+  CAMLparam1(vData);
+  CAMLlocal1(vStr);
+
+  vStr = caml_copy_string("Unimplemented");
+
+  CAMLreturn(vStr);
+}
+
+// Again, this function doesn't make much sense on Windows, so it's a noop.
+CAMLprim value oni2_SparkleDebugLog(value vData) {
+  CAMLparam1(vData);
+  
+  CAMLreturn(Val_unit);
+}
+
+CAMLprim value oni2_SparkleSetFeedURL(value vUpdater, value vUrlStr) {
+  CAMLparam2(vUpdater, vUrlStr);
+
+  char *urlStr = malloc(caml_string_length(vUrlStr) * sizeof(char));
```

Comment:
I believe this would actually need to be `caml_string_length(vUrlStr) + 1`, to leave an extra byte for the null-terminating character.

At least, the examples seem to do this too:
```
  /* Copy the string argument to a C string, allocated outside the
     OCaml heap. */
  name = caml_stat_alloc(caml_string_length(vname) + 1);
  name = caml_strdup(String_val(vname));
```
http://caml.inria.fr/pub/distrib/ocaml-4.05/ocaml-4.05-refman.html

I believe we could just use those same functions here

Reply:
Interesting! I just changed it to this in 3302915, but the way those functions work is a little unclear to me -- what is the point of allocating the pointer in the first line, only to discard it in the second? I know you didn't write the example, I'm just confused haha.

Here I see a line with the exact same comment, but done in a more understandable (at least to me!) one-line solution: https://caml.inria.fr/pub/docs/manual-ocaml/intfc.html
```c
 /* Copy the string argument to a C string, allocated outside the
     OCaml heap. */
  name = caml_stat_strdup(String_val(vname));
```

I wonder which is "better" -- the link you provided is for OCaml 4.05, so it might be out of date? I'll look into it and see if anything's changed!

Reply:
Looks like it may be out of date, actually -- the OCaml 4.10 manual has the same line as the one in the FFI docs:
http://caml.inria.fr/pub/distrib/ocaml-4.10/ocaml-4.10-refman.html

I'll update it to that one!

Reply:
Fixed in ce55580!